RFC-0012-materialized-views.md

```diff
@@ -0,0 +1,1029 @@
+
+* Existing Hive Materialized Views implementation in presto-hive module
+* Proposed Iceberg Materialized Views support
+
+## Summary
+
+This RFC proposes a redesign of Presto's Materialized Views architecture to move from a Hive-centric, analysis-phase-heavy implementation to a connector-agnostic, planning-phase-driven architecture that better supports modern table formats like Apache Iceberg. The design emphasizes a minimal SPI with advanced features like stitching handled through the existing `ConnectorPlanOptimizerProvider` mechanism rather than complex SPI methods.
+
+## Background
+
+The current Materialized Views implementation in Presto was designed specifically for Hive and has several architectural limitations that prevent modern connectors from leveraging their advanced capabilities (detailed in the Problems section below). Modern table formats like Apache Iceberg provide advanced features (hidden partitioning, snapshot isolation, incremental scans) that could enable much more efficient materialized view implementations, but the current architecture cannot leverage these capabilities.
+
+### Goals
+
+* Create a connector-agnostic materialized views framework
+* Move complex MV logic from analysis phase to planning phase
+* Enable connectors to implement efficient, format-specific refresh strategies
+* Support incremental refresh patterns natively
+* Provide a complete Iceberg connector implementation leveraging snapshot-based change tracking
+
+### Non-goals
+
+* Backward compatibility with existing Hive MV implementation (it's undocumented and unused)
+* Supporting materialized views across different connectors (cross-connector MVs)
+* Automatic view selection/routing for arbitrary queries
+
+## Current Implementation and How It Works
+
+The existing materialized views implementation is located primarily in the presto-hive module and follows this flow:
+
+### 1. Metadata Storage
+- MVs are stored as Hive tables with special properties
+- View definition stored in table properties as serialized SQL
+- Base table references tracked in metadata
+
+### 2. Query Analysis Phase
+During analysis, the system:
+- Parses the MV definition from metadata
+- Constructs the refresh query by analyzing the view definition
+- Determines partitioning alignment between MV and base tables
+- Calculates freshness based on partition modification times
+
+### 3. Refresh Execution (Hive-Specific)
+- Full refresh: Drops and recreates the entire MV table
+- Partial refresh: Attempts to update only modified partitions (detected via Hive partition modification timestamps)
+- Uses Hive's partition replacement semantics
+- **Limitation**: Can only detect partition-level changes through Hive metastore's `transient_lastDdlTime`
+
+### 4. Query Rewriting
+- Simple pattern matching to detect if a query can use an MV
+- Limited to exact matches or simple projections
+- No cost-based selection between multiple MVs
```

Comment:
I think this is also a limitation of the current design since the analyzer is still doing the replacement

Reply:
But the problem remains when we would do "view replacement", we would need to create some canonicalized AST to replace.

Reply:
I believe the new design will handle this better, or at least I can see a high level path on how we can get that to work.  Connectors return a list of possible replacement candidates via the existing `ConnectorMetadata#getReferencedMaterializedViews` for a given table during analysis.  We put this in the `MaterializedViewNode` AST, and plan each of all of the candidates.  At this point I believe we have all of the tools necessary to leverage some of the techniques that have been written about, for example the technique Calcite uses.